node-red-contrib-ms-language/node-luis.js

```diff
@@ -1,6 +1,113 @@
 const request =    require('request-promise');
 const helper     = require('node-red-viseo-helper');
 
+const extractEntities = (prediction) => {
+    const entities = [];
+    const compositeEntities = [];
+
+    Object.values(prediction.entities.$instance)
+    .reduce((result, entity) => result.concat(...entity), []) // flat entity array
+    .forEach(entity => {
+        if (!entity) return;
+        if (entity.modelType === 'Composite Entity Extractor') {
+            compositeEntities.push({
+                parentType: entity.type,
+                value: entity.text,
+                children: []
+            });
+            Object.values(prediction.entities[entity.type][0].$instance).forEach(child => {
+                compositeEntities[compositeEntities.length - 1].children.push({
+                  type: child[0].type,
+                  value: child[0].text
+                });
+                entities.push({
+                  type: child[0].type,
+                  entity: child[0].text,
+                  startIndex: child[0].startIndex,
+                  endIndex: child[0].startIndex + child[0].length,
+                  score: entity.score
+                })
+              });
+        } else {
+            entities.push({
+                type: entity.type,
+                entity: entity.text,
+                startIndex: entity.startIndex,
+                endIndex: entity.startIndex + entity.length, // add end index
+                resolution: {
+                  values: response.prediction.entities[entity.type.replace('builtin.', '')].reduce((result, e) => {
+                    if (Array.isArray(e)) {
+                      return result.concat(...e);
+                    }
+                    result.push(e);
+                    return result;
+                  }, [])
+                }
+            });
+        }
+    });
+
+    return {
+        entities,
+        compositeEntities
+    };
+};
+
+const asyncGetPrediction = async function(node, text) {
+    let entityObject = {
+        entities: [],
+        compositeEntities: []
+    };
+
+    const response = await request({
+        uri: `${node.endpoint}${encodeURIComponent(text)}`,
+        method: 'GET',
+        headers: {
+            'Ocp-Apim-Subscription-Key': node.subKey,
+            ...(node.spellCheckKey && {'mkt-bing-spell-check-key': node.spellCheckKey})
+        },
+        json: true
+    });
+
+    // Merge additional metadate in the $instance object to entities
+    if (response.prediction.entities.$instance) {
+        entityObject = extractEntities(response.prediction);
```

Comment:
Here you are overriding the entityObject which was declared with empty arrays line 57. It would be better to move the condition inside the extractEntities function and use this function to always return an object with the right format.

Reply:
There is a typo in the comment, it should be "metadata"